File: server/models/Timetable.js
const mongoose = require("mongoose");

const activitySchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true,
  },
  time: {
    type: String,
    required: true,
  },
  category: {
    type: String,
    required: true,
  },
});

const dailyProgressSchema = new mongoose.Schema({
  activity: {
    type: activitySchema,
    required: true,
  },
  dailyStatus: {
    type: [Boolean],
    default: [false, false, false, false, false, false, false],
    validate: {
      validator: function (arr) {
        return arr.length === 7;
      },
      message: "Daily status must contain exactly 7 values",
    },
  },
  completionRate: {
    type: Number,
    default: 0,
  },
});

const weekSchema = new mongoose.Schema({
  weekStartDate: {
    type: Date,
    required: true,
  },
  weekEndDate: {
    type: Date,
    required: true,
  },
  activities: [dailyProgressSchema],
  overallCompletionRate: {
    type: Number,
    default: 0,
  },
  notes: {
    type: String,
    trim: true,
  },
});

const timetableSchema = new mongoose.Schema(
  {
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: true,
    },
    name: {
      type: String,
      required: true,
      trim: true,
      default: "Default Timetable",
    },
    description: {
      type: String,
      trim: true,
    },
    isActive: {
      type: Boolean,
      default: true,
    },
    currentWeek: {
      type: weekSchema,
      required: true,
    },
    history: [weekSchema],
    defaultActivities: [activitySchema],
  },
  { timestamps: true }
);

timetableSchema.pre("save", function (next) {
  if (this.currentWeek && this.currentWeek.activities) {
    let totalPossible = this.currentWeek.activities.length * 7;
    let totalCompleted = 0;

    this.currentWeek.activities.forEach((activity) => {
      const completed = activity.dailyStatus.filter(Boolean).length;
      activity.completionRate = (completed / 7) * 100;
      totalCompleted += completed;
    });

    this.currentWeek.overallCompletionRate =
      totalPossible > 0 ? (totalCompleted / totalPossible) * 100 : 0;
  }

  this.history.forEach((week) => {
    let totalPossible = week.activities.length * 7;
    let totalCompleted = 0;

    week.activities.forEach((activity) => {
      const completed = activity.dailyStatus.filter(Boolean).length;
      activity.completionRate = (completed / 7) * 100;
      totalCompleted += completed;
    });

    week.overallCompletionRate =
      totalPossible > 0 ? (totalCompleted / totalPossible) * 100 : 0;
  });

  next();
});

timetableSchema.methods.startNewWeek = async function () {
  console.log("Starting new week...");

  if (
    this.currentWeek &&
    this.currentWeek.activities &&
    this.currentWeek.activities.length > 0
  ) {
    console.log("Moving current week to history:", {
      startDate: this.currentWeek.weekStartDate,
      endDate: this.currentWeek.weekEndDate,
      activities: this.currentWeek.activities.length,
    });
    this.history.push(this.currentWeek);
  }

  const now = new Date();
  const currentDay = now.getDay();

  const monday = new Date(now);

  const daysToSubtract = currentDay === 0 ? 6 : currentDay - 1;
  monday.setDate(now.getDate() - daysToSubtract);
  monday.setHours(0, 0, 0, 0);

  const sunday = new Date(monday);
  sunday.setDate(monday.getDate() + 6);
  sunday.setHours(23, 59, 59, 999);

  console.log("New week dates:", {
    monday: monday.toISOString(),
    sunday: sunday.toISOString(),
    now: now.toISOString(),
  });

  this.currentWeek = {
    weekStartDate: monday,
    weekEndDate: sunday,
    activities: this.defaultActivities.map((activity) => ({
      activity: activity,
      dailyStatus: [false, false, false, false, false, false, false],
      completionRate: 0,
    })),
    overallCompletionRate: 0,
  };

  console.log(
    "New week created with activities:",
    this.currentWeek.activities.length
  );

  return this;
};

timetableSchema.index({ user: 1, name: 1 }, { unique: true });

module.exports = mongoose.model("Timetable", timetableSchema);

File: backend-go/internal/models/timetable.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Activity struct {
	Name     string `bson:"name" json:"name"`
	Time     string `bson:"time" json:"time"`
	Category string `bson:"category" json:"category"`
}

type DailyProgress struct {
	ID             primitive.ObjectID `bson:"_id,omitempty" json:"_id"` // Added ID
	Activity       Activity           `bson:"activity" json:"activity"`
	DailyStatus    []bool             `bson:"dailyStatus" json:"dailyStatus"` // array of 7 bools
	CompletionRate float64            `bson:"completionRate" json:"completionRate"`
}

type Week struct {
	WeekStartDate         time.Time       `bson:"weekStartDate" json:"weekStartDate"`
	WeekEndDate           time.Time       `bson:"weekEndDate" json:"weekEndDate"`
	Activities            []DailyProgress `bson:"activities" json:"activities"`
	OverallCompletionRate float64         `bson:"overallCompletionRate" json:"overallCompletionRate"`
	Notes                 string          `bson:"notes,omitempty" json:"notes,omitempty"`
}

type Timetable struct {
	ID                primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	User              primitive.ObjectID `bson:"user" json:"user"`
	Name              string             `bson:"name" json:"name"`
	Description       string             `bson:"description,omitempty" json:"description,omitempty"`
	IsActive          bool               `bson:"isActive" json:"isActive"`
	CurrentWeek       Week               `bson:"currentWeek" json:"currentWeek"`
	History           []Week             `bson:"history" json:"history"`
	DefaultActivities []Activity         `bson:"defaultActivities" json:"defaultActivities"`
	CreatedAt         time.Time          `bson:"createdAt" json:"createdAt"`
	UpdatedAt         time.Time          `bson:"updatedAt" json:"updatedAt"`
}
